drivers/sensor/default_rtio_sensor.c

```diff
@@ -0,0 +1,426 @@
+	uint32_t min_buf_len = compute_min_buf_len(num_channels, num_output_samples);
+	uint64_t timestamp_ns = k_ticks_to_ns_floor64(k_uptime_ticks());
+	int rc = sensor_sample_fetch(dev);
+	uint8_t *buf;
+	uint32_t buf_len;
+
+	/* Check that the fetch succeeded */
+	if (rc != 0) {
+		LOG_ERR("Failed to fetch samples");
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return;
+	}
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return;
+	}
+
+	/* Set the timestamp and num_channels */
+	struct sensor_data_generic_header *header = (struct sensor_data_generic_header *)buf;
+
+	header->timestamp_ns = timestamp_ns;
+	header->num_channels = num_channels;
+
+	uint8_t current_channel_index = 0;
+	q31_t *q = (q31_t *)(buf + sizeof(struct sensor_data_generic_header) +
+			     num_channels * sizeof(struct sensor_data_generic_channel));
+
+	for (size_t i = 0; i < num_channels; ++i) {
+		struct sensor_value value[3];
+		int num_samples = SENSOR_CHANNEL_3_AXIS(channels[i]) ? 3 : 1;
+
+		/* Get the current channel requested by the user */
+		rc = sensor_channel_get(dev, channels[i], value);
+		if (rc != 0) {
+			LOG_INF("Failed to get channel %d, skipping", channels[i]);
+			header->channel_info[i].channel = SENSOR_CHAN_MAX;
+			current_channel_index += num_samples;
+			continue;
+		}
+
+		/* Get the largest absolute value reading to set the scale for the channel */
+		uint32_t header_scale = 0;
+
+		for (int sample = 0; sample < num_samples; ++sample) {
+			uint32_t scale = (uint32_t)llabs((int64_t)value[sample].val1) + 1;
+
+			header_scale = MAX(header_scale, scale);
+		}
+		header->channel_info[i].channel = channels[i];
+		header->channel_info[i].shift = ilog2(header_scale - 1) + 1;
+		LOG_DBG("header->channel_info[%zu](channel=%d, shift=%d, header_scale=%u)", i,
+			channels[i], ilog2(header_scale - 1) + 1, header_scale);
+
+		/*
+		 * Spread the q31 values. This is needed because some channels are 3D. If
+		 * the user specified one of those then num_samples will be 3; and we need to
+		 * produce 3 separate readings.
+		 */
+		for (int sample = 0; sample < num_samples; ++sample) {
+			/* Check if the channel is already in the buffer */
+			int prev_computed_value_idx =
+				check_header_contains_channel(header, channels[i], sample);
+
+			if (prev_computed_value_idx >= 0 &&
+			    prev_computed_value_idx != current_channel_index + sample) {
+				LOG_DBG("value[%d] previously computed at q[%d]@%p", sample,
+					prev_computed_value_idx,
+					(void *)&q[prev_computed_value_idx]);
+				q[current_channel_index + sample] = q[prev_computed_value_idx];
+				continue;
+			}
+
+			/* Convert the value to micro-units */
+			int64_t value_u =
+				value[sample].val1 * INT64_C(1000000) + value[sample].val2;
```

Comment:
can you use `sensor_value_to_micro` here?

Reply:
Sure could